File: frontend/src/lib/jar.test.ts
import { describe, it, expect } from "vitest";
import { parseJarFilename } from "./jar";

describe("parseJarFilename", () => {
  const cases = [
    {
      file: "sodium-fabric-mc1.20.1-0.4.10.jar",
      slug: "sodium",
      id: "sodium",
      version: "0.4.10",
      mcVersion: "1.20.1",
      loader: "fabric",
    },
    {
      file: "jei-1.20.1-forge-15.2.0.27.jar",
      slug: "jei",
      id: "jei",
      version: "15.2.0.27",
      mcVersion: "1.20.1",
      loader: "forge",
    },
    {
      file: "fabric-api-0.86.1+1.20.1.jar",
      slug: "fabric-api",
      id: "fabric",
      version: "0.86.1",
      mcVersion: "1.20.1",
      loader: "fabric",
    },
    {
      file: "awesome-mod-1.2.3-beta.jar",
      slug: "awesome-mod",
      id: "awesome",
      version: "1.2.3",
      channel: "beta",
    },
    {
      file: "example-rc-v2.0.0.jar",
      slug: "example",
      id: "example",
      version: "2.0.0",
      channel: "rc",
    },
  ];
  for (const c of cases) {
    it(c.file, () => {
      const got = parseJarFilename(c.file);
      expect(got.slug).toBe(c.slug);
      expect(got.id).toBe(c.id);
      expect(got.version).toBe(c.version);
      expect(got.mcVersion).toBe(c.mcVersion ?? "");
      expect(got.loader).toBe(c.loader ?? "");
      expect(got.channel).toBe(c.channel ?? "");
    });
  }
});

File: internal/handlers/jar_test.go
package handlers

import "testing"

func TestParseJarFilename(t *testing.T) {
	tests := []struct {
		name string
		file string
		want jarMeta
	}{
		{
			name: "sodium",
			file: "sodium-fabric-mc1.20.1-0.4.10.jar",
			want: jarMeta{Slug: "sodium", ID: "sodium", Version: "0.4.10", MCVersion: "1.20.1", Loader: "fabric"},
		},
		{
			name: "jei",
			file: "jei-1.20.1-forge-15.2.0.27.jar",
			want: jarMeta{Slug: "jei", ID: "jei", Version: "15.2.0.27", MCVersion: "1.20.1", Loader: "forge"},
		},
		{
			name: "fabric-api",
			file: "fabric-api-0.86.1+1.20.1.jar",
			want: jarMeta{Slug: "fabric-api", ID: "fabric", Version: "0.86.1", MCVersion: "1.20.1", Loader: "fabric"},
		},
		{
			name: "beta channel",
			file: "awesome-mod-1.2.3-beta.jar",
			want: jarMeta{Slug: "awesome-mod", ID: "awesome", Version: "1.2.3", Channel: "beta"},
		},
		{
			name: "rc channel",
			file: "example-rc-v2.0.0.jar",
			want: jarMeta{Slug: "example", ID: "example", Version: "2.0.0", Channel: "rc"},
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := parseJarFilename(tt.file)
			if got.Slug != tt.want.Slug || got.ID != tt.want.ID || got.Version != tt.want.Version || got.MCVersion != tt.want.MCVersion || got.Loader != tt.want.Loader || got.Channel != tt.want.Channel {
				t.Errorf("parseJarFilename(%q) = %+v, want %+v", tt.file, got, tt.want)
			}
		})
	}
}
